### PR TITLE
[sled-agent] Add tests for service management, fix prefix-matching bug

### PR DESCRIPTION
**Pull request by smklein**
## Background

@leftwo helped identify an issue where the ServiceManager attempts to create a service which already exists because it failed to identify that one was already running, based on the name. This resulted in a "DLPI link error", as the zone management logic tried to re-allocate an in-use IP address.

## What this patch does

- Refactors `ServiceManager` to be slightly easier to test
- Fixes the prefix matching issue (it now checks for the existence of "oxz_my_svc" for a service named "my_svc", which is what the zone is actually named).
- Adds a regression test for this issue, as well as tests for a handful of other basic behaviors.

### CODE REVIEW

File: sled-agent/src/services.rs [hunk 1]
```diff
@@ -105,7 +129,9 @@ impl ServiceManager {
         for service in services {
             // Before we bother allocating anything for this request, check if
             // this service has already been created.
-            if existing_zones.iter().any(|z| z.name() == service.name) {
+            let expected_zone_name =
+                InstalledZone::get_zone_name(&service.name, None);
+            if existing_zones.iter().any(|z| z.name() == expected_zone_name) {
```
Comment (smklein) [inline sled-agent/src/services.rs:134]: This is the actual bug fix

File: sled-agent/src/services.rs [hunk 2]
```diff
@@ -203,12 +233,227 @@ impl ServiceManager {
...
+        let mgr = ServiceManager::new(
+            log,
+            Some(PhysicalLink(EXPECTED_LINK_NAME.to_string())),
+            Some(config),
+        )
+        .await
+        .unwrap();
+
+        ensure_new_service(&mgr).await;
+        drop_service_manager(mgr);
```
Comment (bnaecker) [inline sled-agent/src/services.rs:362]: I think you need the code to indicate the test was successful: `logctx.cleanup_successful()`.

File: sled-agent/src/illumos/dladm.rs [hunk 3]
```diff
@@ -27,7 +27,7 @@ pub enum Error {
 }
 
 /// The name of a physical datalink.
-#[derive(Debug, Clone, Deserialize, Serialize)]
+#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
```
Comment (bnaecker) [inline sled-agent/src/illumos/dladm.rs:30]: Not related to this PR, but I think this type should probably not be `Clone`.

**Review by bnaecker — APPROVED** [review]

File: sled-agent/src/illumos/dladm.rs (hunk 3 above)
Comment (smklein) [reply sled-agent/src/illumos/dladm.rs:30]: I kinda disagree - this type doesn't represent the physical link itself, it's just a strongly-typed name. Shouldn't names be cloneable?

File: sled-agent/src/services.rs (hunk 2 above)
Comment (smklein) [reply sled-agent/src/services.rs:362]: ugh, I always forget this! Thanks, added.

File: sled-agent/src/illumos/dladm.rs (hunk 3 above)
Comment (bnaecker) [reply sled-agent/src/illumos/dladm.rs:30]: That's fair -- I think I'm conflating the case where it's a created device like a VNIC, which it clearly differs from. Ignore :)

**Pull request merged**